<!-- Written by: Yun Yan (https://github.com/Puriney) -->

# Preprocessing


All the following operations are using the tutorial folder path as the working directory. 

``` shell
cd PATH_IN_YOUR_COMPUTER/wellDA-seq/tutorial
```

## 1. Preprocessing the scATAC modality data

Input: 
1. The pair-end FASTQ files for the scATAC modality. 
    - <kbd>fq/DCIS66T_chip2/atac/DCIS66T_chip2.demplxed.PE1.fastq.gz</kbd>
    - <kbd>fq/DCIS66T_chip2/atac/DCIS66T_chip2.demplxed.PE2.fastq.gz</kbd>
2. The text file recording the wells with cells dispensed. 
    - <kbd>wafergen/135581_insue_WellList.TXT</kbd>

With FASTQ files as input, we use scATAC-pro to map reads to genome and generates the ATAC
 fragment file of Tn5 insertions. The ATAC fragment file is the common format of inputs of other downstream analysis tools, including ArchR and Signac. 

Expected output: a Signac object.

### 1.1. Generating scATAC fragment file by scATAC-pro

Here is an example of a particular read in the FASTQ file. It is from the cell 'A13_A10'. With this format, scATAC-pro will assign data to each single cell.

``` shell
gunzip -c DCIS66T_chip2.demplxed.PE1.fastq.gz | head -8
```

```
@A13_A10:VH00219:49:AACYKCYM5:1:1101:20674:7759 1:N:0:TAAGGCGAATC+CGATACACTCA
CTAAAGGGTTGTAGCTGTGAAGGAATAATGGGCGGAGTGGTGGCACACACGGCATT
+
C;C-CC-CC-CCC-;CCCCCC;CCCCCCCCC-CC;CC;CCCC;;CCCCCCCCCCCC
@A13_A10:VH00219:49:AACYKCYM5:1:1101:23420:27239 1:N:0:TAAGGCGAATC+CGATACACTCA
AGGCTGGGAAGCCCAAGACCAAAGTACCAGCAGATTTGGTGTCTGGTGAGGGGCTG
+
CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC
```


Create a configuration file for scATAC-pro. To run each sample, scATAC-pro requires a separate configuration file, which specifies the prerequisite files, such as path to the aligner, the genome index, and the genome version. Here, the config file of DCIS66T_chip2 is 'DCIS66T_chip2.config.txt' in the 'scripts' folder. For example, here the reads are mapped to the human genome assembly hg19 by using the aligner bowtie2.

Run scATAC-pro by using the shell script 's2.scatacpro.sh' in the 'scripts' folder.

``` shell
sh scripts/s2.scatacpro.sh DCIS66T_chip2 "atac"
```

Below are the expected results:

```
.
├── DCIS66T_chip2.fragments.tsv.len.gz
├── DCIS66T_chip2.MACS2.qc_per_barcode.txt
├── DCIS66T_chip2.MappingStats
├── cell_barcodes.MappingStats
├── DCIS66T_chip2.fragments.tsv.gz      <---- fragment file !!!
├── DCIS66T_chip2.fragments.tsv.gz.tbi      
└── scATAC-pro_report_DCIS66T_chip2.html
```

The first 10 lines of the fragment, which is the standard format as described by 10X ([link](https://www.10xgenomics.com/support/software/cell-ranger-arc/latest/analysis/fragments-file)):

```shell
gunzip -c DCIS66T_chip2.fragments.tsv.gz | head -10
```


```
chr1    10119   10239   L14_C7  1
chr1    10443   10544   M16_A9  2
chr1    10472   10566   C16_D10 3
chr1    10495   10612   C17_K8  4
chr1    10495   10595   A15_O8  4
chr1    10522   10600   P15_L7  3
chr1    10558   10627   N14_D10 4
chr1    13041   13109   I14_P7  4
chr1    13452   13542   L13_L7  3
chr1    13458   13529   J16_D10 5
```

### 1.2. Quality-control by ArchR

The quality of single cell is evaluated based on TSS enrichment score and the number of unique fragments, which are computed using ArchR. 

The initial quality-control on the scATAC data was performed mainly based on the TSS enrichment score (no less than 8) and the number of unique fragments (no less than 1000) of each cell. 


<details>
<summary> What is the appropriate cutoff for TSS enrichment score? </summary>
Different cutoff is suggested depending on species and the genome assembly version. We chose 8 for hg19 (see <a href="https://doi.org/10.1038%2Fs41587-019-0206-z">Satpathy et al 2019</a>). Further read the instructions from ENCODE: https://www.encodeproject.org/atac-seq/. 
</details>

<details>
<summary> Why using ArchR to calcualte the TSS enrichment score? </summary>
We find the TSS score calcualted by ArchR is morely likely in the value range of the ENCODE instruction, whereas the Signac calculation is not. 
</details>

<details>
<summary> What is the appropriate cutoff for fragment number? </summary>
It depends on the specific data because of many factors, for example, the sequencing depth. In most cases, 1000 is a reasonable number to remove the empty or low quality cell (see <a href="https://doi.org/10.1038%2Fs41587-019-0206-z">Satpathy et al 2019</a>).
</details>


``` bash session
s=DCIS66T_chip2
Rscript scripts/create_archr.public.R \
    ${s}_atac \
    ${s}/atac/output/summary/${s}.fragments.tsv.gz 
```

| Basic QC                                                                                                                                    | Fragment size distribution                                                                                                          | Doublet detection in silico                                                                                                                        |
| ------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------- |
| <img src="https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/atac_QC_basic.png?raw=true" alt="txt" width="400"> | <img src="https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/atac_QC_tri_archr.png?raw=true" width=400> | <img src="https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/atac_QC_doublet_init.png?raw=true" alt="txt" width="400"> |

In addition, a typical good ATAC experiment should show the periodical enrichment of the fragment size, corresponding to the nucleosome-free regions and mono-/di-/tri-nucleosomes ([Yan, F., Powell, D.R., Curtis, D.J. et al. 2020](https://doi.org/10.1186/s13059-020-1929-3))

Last, we used the ArchR to perform the basic [doublet detection in silico](https://www.archrproject.com/bookdown/filtering-doublets-from-an-archrproject.html). Note that our doublet detection dose not only reply on this. In our best practice, the further clustering and the co-assay data will identify some more doublets. 


### 1.3. Creating Signac object for analysis

This script will create a Signac object to prepare the scATAC data analysis. It will 1) filter cells based on the TSS enrichment score and the number of unique fragments that are measured by ArchR, 2) keep cells with the valid dispensing record, and 3) perform the basic analysis including clustering and UMAP visualization to give a quick overview of the possible clusters.  

``` shell
s=DCIS66T_chip2
Rscript scripts/create_signac.public.R 
    ${s}_atac \
    $s/atac/output/summary/$s.fragments.tsv.gz \
    8 "1e3" "1e5" 0 1 
```

The last 4 parameters are used to control how to filter cells: 
1) define the min TSS score per cell. Default: 8. 
2) define the min number of fragments per cell. Default: 1,000. 
3) define the max number of fragments per cell. Default: 10,000. If the fragment number is too high in a cell, this cell could be very likely to be a doublet -- should be removed from the downstream analysis.  
4) whether binarize the data. 0=non-binarized, 1=binarized. In the non-cancer and normal samples, this should be binarized. In the cancer samples, we recommend not binarizing the data. 
5) whether use the QC results of ArchR. Default: 1. 0=No, 1=Yes. 

| Physical dispensing cells                                                                                                                                     | Basic UMAP                                                                                                                                           |
| ------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------- |
| <img src="https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/atac.wafargen_physical_dispense.png?raw=true" alt="txt" width="400"> | <img src="https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/atac_overview_umap_snn.png?raw=true" alt="txt" width="400"> |


We can visualize the record of dispensing cells on the wafergen chip. The majority of the empty wells (white) of course have no cells passing QC. The majority of the dispensed cells (blue) successfully pass the ATAC QC. Due to technical errors, it is not surprising to find that the cells in some dispensed wells (grey) fail the ATAC QC. These false-negative wells (grey) form some physical patterns on the chip, mainly because the dispensing tip of wafergen does not technically work well in that particular rows or columns. 

<details>
<summary> What if there are a lot orange wells? </summary>
If there are many false-positive cells (orange) in your experiment, it indicates a fatal error -- the sequenced data contains cells that are not dispensed at all. In this situation, it could arise from several reasons: 1) dispensing record of the wafergen experiment is wrong; 2) demultiplexing FASTQ is wrong; 3) visualization is wrong. 

</details>

In this step, this script will also generate a UMAP for a quick visualization of cell clusters. Here, we expect there are some epithelial cell populations and some immune/stromal cell populations, so the data looks good for now. 

Now, the Signac object is ready for further downstream processing and analysis. 

## 2. Preprocessing the scDNA modality data

Input: 
1. The folder saving the de-multiplexed FASTQ files for the scDNA modality. 
    - <kbd>fq/DCIS66T_chip2/dna</kbd>
2. The text file recording the wells with cells dispensed. 
    - <kbd>wafergen/135581_insue_WellList.TXT</kbd>

Expected output: A Copykit object. 

### 2.1. Generating scCNA matrix and quality control by CNA-pipeline

The CNA-pipeline will use call the copy number alterations at a ~220kb resolution. 

``` shell
for s in DCIS66T_chip2
do 
    echo $s
    resdir="./rds/${s}_dna"
    mkdir -p $resdir
    cd $resdir
    ls fq/$s/dna/*/*.gz > $resdir/sample_list.txt
    <PATH_IN_YOUR_COMPUTER>/CNA_pipeline_v1.4/run_CNA.sh \
        --fastq $resdir/sample_list.txt \
        --res 200 -o $resdir/res_200_k_reg
done
```

The result of the pipeline includes the copy number alteration matrix and a html file showing the basic QC and the heatmap showing the log2 CNA ratio. 

<p align="center">
    <img src="https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/dna_snapshot_CNApipeline.png?raw=true" alt="txt" width="400" >
</p>


The full list of the outputs of the pipeline is: 
```
./res_200_k_reg/final_result
├── Figures
├── IGV_longs
├── IGV_shorts
├── ratio_plots
├── ratio_plots_CN
├── sample_cna_dashboard.html <----- Html file showing basic QC
├── uber.sample.bin.txt       <----- CNA counts for bins
├── uber.sample.ratio.txt     <----- CNA ratio for bins
└── uber.sample.seg.txt       <----- CNA counts for segments
```


### 2.2. Creating copykit object for analysis

This script will create the Copykit object which contains the single-cell copy number matrix data. 

``` shell
for s in DCIS66T_chip2_dna 
do 
    Rscript scripts/create_copykit.conventional.R \
      /rds/$s/res_200_k_reg/final_result \
      /rds/$s $s &
done
```

Using the wafergen dispensing file, this script will keep the scDNA data of the dispensed cells. Basic clustering and visualization are also included to give a quick overview of the subclones. 

```shell
for s in DCIS66T_chip2_dna 
do
    Rscript scripts/create_copykit.filter_qc_cells.R \
      rds/$s/res_200_k_reg/final_result \
      rds/$s \
      $s 'tissue' &
done
```

The last parameter controls the type of the input sample. It could be either 'cell_line' or 'tissue'. If 'cell_line', the script will use depend on the Copykit's function `findAneuploidCells` to filter out [non-aneuploid cells](https://navinlabcode.github.io/CopyKit-UserGuide/qc-module.html#findaneuploidcells). If 'tissue', the script will not perform this selection because it might consider the cancer progenitor cells as non-aneuploid cells, given the small CNA events of the cancer progenitor cells. 


| Physical dispensing cells                                                                                             | Basic UMAP                                                                                                                          | Heatmap                                                                                                                                  |
| --------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------- |
| ![dispense](https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/dna_dispense.png?raw=true) | ![umap](https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/dna.clean.plotUmap.superclones.png?raw=true) | ![heatmap](https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/dna.clean.plotHeatmap.clones.004.png?raw=true) |


The DNA modality is also good because the majority of dispensed cells (blue) pass the basic DNA QC. There are no false-positive cells (orange) at all. 

In this step, we performed the basic clustering on the scDNA data and resulted in the superclones, representing the major populations with distinct CNAs profiles. In most cases, it require further clustering to resolve the subclones. 

The heatmap is also very helpful to visualize the CNA events in details. The majority of cells in this sample are the diploid cells, while there are about 6 superclones harboring CNA events on chr8, chr14 and chr15. 

The next question is: how are the ATAC clusters (i.e., cell types or cell states) and the DNA clusters (super-/sub-clones) are matched? 
