<!-- Written by: Yun Yan (https://github.com/Puriney) -->

# Preprocessing


All the following operations are using the tutorial folder path as the working directory. 

``` bash
cd PATH_IN_YOUR_COMPUTER/wellDA-seq/tutorial
```

## 1. Preprocessing the scATAC modality data

Input: Download the FASTQ files for the scATAC modality. 

- DCIS66T_chip2.demplxed.PE1.fastq.gz
- DCIS66T_chip2.demplxed.PE2.fastq.gz

With FASTQ files as input, we use scATAC-pro to map reads to genome and generates the ATAC
 fragment file of Tn5 insertions. The ATAC fragment file is the common format of inputs of other downstream analysis tools, including ArchR and Signac. 

Expected output: a Signac object.

### 1.1. Generating scATAC fragment file by scATAC-pro

Here is an example of a particular read in the FASTQ file. It is from the cell 'A13_A10'. With this format, scATAC-pro will assign data to each single cell.

``` bash
zcat DCIS66T_chip2.demplxed.PE1.fastq.gz | head -8
```

```
@A13_A10:VH00219:49:AACYKCYM5:1:1101:20674:7759 1:N:0:TAAGGCGAATC+CGATACACTCA
CTAAAGGGTTGTAGCTGTGAAGGAATAATGGGCGGAGTGGTGGCACACACGGCATT
+
C;C-CC-CC-CCC-;CCCCCC;CCCCCCCCC-CC;CC;CCCC;;CCCCCCCCCCCC
@A13_A10:VH00219:49:AACYKCYM5:1:1101:23420:27239 1:N:0:TAAGGCGAATC+CGATACACTCA
AGGCTGGGAAGCCCAAGACCAAAGTACCAGCAGATTTGGTGTCTGGTGAGGGGCTG
+
CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC
```


Create a configuration file for scATAC-pro. To run each sample, scATAC-pro requires a separate configuration file, which specifies the prerequisite files, such as path to the aligner, the genome index, and the genome version. Here, the config file of DCIS66T_chip2 is 'DCIS66T_chip2.config.txt' in the 'scripts' folder. For example, here the reads are mapped to the human genome assembly hg19 by using the aligner bowtie2.

Run scATAC-pro by using the bash script 's2.scatacpro.sh' in the 'scripts' folder.

``` bash
sh scripts/s2.scatacpro.sh DCIS66T_chip2 "atac"
```

Below are the expected results:

```
.
├── DCIS66T_chip2.fragments.tsv.len.gz
├── DCIS66T_chip2.MACS2.qc_per_barcode.txt
├── DCIS66T_chip2.MappingStats
├── cell_barcodes.MappingStats
├── DCIS66T_chip2.fragments.tsv.gz      <---- fragment file !!!
├── DCIS66T_chip2.fragments.tsv.gz.tbi      
└── scATAC-pro_report_DCIS66T_chip2.html
```

The first 10 lines of the fragment, which is the standard format as described by 10X ([link](https://www.10xgenomics.com/support/software/cell-ranger-arc/latest/analysis/fragments-file)):

```bash
zcat DCIS66T_chip2.fragments.tsv.gz | head -10
```


```
chr1    10119   10239   L14_C7  1
chr1    10443   10544   M16_A9  2
chr1    10472   10566   C16_D10 3
chr1    10495   10612   C17_K8  4
chr1    10495   10595   A15_O8  4
chr1    10522   10600   P15_L7  3
chr1    10558   10627   N14_D10 4
chr1    13041   13109   I14_P7  4
chr1    13452   13542   L13_L7  3
chr1    13458   13529   J16_D10 5
```

### 1.2. Quality-control by ArchR

The intial quality-control on the scATAC data was performed mainly based on the TSS enrichment score (no less than 8) and the number of unique fragments (no less than 1000) of each cell. 


``` bash
s=DCIS66T_chip2
Rscript scripts/create_archr.public.R \
    ${s}_atac \
    ${s}/atac/output/summary/${s}.fragments.tsv.gz 
```

!!image xxx!!

<details>
<summary> What is the appropriate cutoff for TSS enrichment score? </summary>
Different cutoff is suggested depending on the genome assembly and species. Further read the instructions from ENCODE: https://www.encodeproject.org/atac-seq/. 
</details>

<details>
<summary> What is the appropriate cutoff for fragment number? </summary>
It depends on the specific data because of many factors, for example, the sequencing depth. In most cases, 1000 is a reasonable number to remove the empty or low quality cell (see <a href="https://doi.org/10.1038%2Fs41587-019-0206-z">Satpathy et al 2019</a>).
</details>



### 1.3. Creating Signac object for analysis







## 2. Preprocessing the scDNA modality data

### 2.1. Generating scCNA matrix and quality control by CNA-pipeline

### 2.2. Creating copykit object for analysis

