# Annotate cell types using the scATAC modality data

**Full R script**: <kbd>scripts/annotate_conda.pipe.R</kbd> ([link](https://github.com/navinlabcode/wellDA-seq/tree/main/tutorial/scripts/annotate_conda.pipe.R))

**Rational**: the annotation of ATAC clusters is based on their top deferentially expressed genes (DEGs) inferred from the ATAC data. During the [pre-processing step](https://github.com/navinlabcode/wellDA-seq/blob/main/tutorial/01.preprocessing.md) for wellDA-seq, the RNA inference from ATAC has been done by using `GeneActivity` of Signac ([link](https://stuartlab.org/signac/articles/pbmc_vignette#create-a-gene-activity-matrix)). 


## 1. Decide a proper clustering resolution

Using the 'clustree' package is helpful to visualize and compare the clustering results of various clustering resolutions to assist determining the proper clustering results. 

```R
clustree::clustree(obja, prefix='peaks_hdbscan_umap_minPts.')
```

| Tree of clusters                                                                                                                                          | A proper clustering resolution                                                                                                                |
| --------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------- |
| <img src="https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/03.clustree_peaks_hdbscan_umap_minPts.png?raw=true" width="400"> | <img src="https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/03.dr.ident_atac_clusters.png?raw=true" width="400"> |


Here, the HDBscan is under-clustering at a resolution of minPt=50 and is probably over-clustering at a resolution of minPt=5. The minPt=20 is a good clustering resolution resulting in a proper number of clusters. 

Additional evaluation we did in the manuscript is to check whether the clusters have similar DEGs. If there are too many clusters sharing the highly similar DEGs, it implies a resolution of over-clustering with redundant clusters. 


## 2. Find top DEGs of each cluster

```R
DefaultAssay(obja) <- 'iRNA'
df_deg0 <- FindAllMarkers(
  obja, min.pct=1/100,
  test.use = 'wilcox',
  assay = 'iRNA', 
  slot = 'data',
  verbose = T
)
df_deg <- df_deg0 %>% dplyr::filter(
  p_val_adj < 0.05, avg_log2FC > log2(1)) %>%
  dplyr::arrange(desc(avg_log2FC))
```

In most cases, the top DEGs will be sufficient to identify the cell types. For example, the cluster-6 is the Luminal hormone responsive (LumHR) epithelial cells because of the marker gene *FOXA1*.  

<img src="https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/03.DEG.png?raw=true" width="400">

## 3. Label transfer if needed

This is a 'difficult' sample having some clusters without a clear cell identity. We picked this sample to show that the real world analysis is not always straightforward and also requires some additional data to assist. Therefore, we computationally integrate this sample with the normal breast single-cell RNA-seq data to seek the possible cell type lineages. Specifically, we performed the label transfer analysis ([link](https://satijalab.org/seurat/articles/seurat5_atacseq_integration_vignette)) by transferring the cell type labels from the HBCA RNA data ([Kumar, T., Nee, K., Wei, R. et al. 2023](https://doi.org/10.1038/s41586-023-06252-9)) to our data. 


```R
or ## Seurat object from HBCA
oa ## Signac object of DCIS66T_chip2
transfer.anchors <- FindTransferAnchors(
  reference = or, query = oa, 
  features = VariableFeatures(or),
  reference.assay = "RNASymbol", 
  query.assay = "iRNA", reduction = "cca")
```

```R
label_transfer_author_cell_type <- TransferData(
  anchorset = transfer.anchors, 
  refdata = or$author_cell_type,
  weight.reduction = oa[["lsi"]], 
  dims = 2:30)
```


| Unbiased clustering                                                                                                                           | After applying label-transfer                                                                                                        |
| --------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------ |
| <img src="https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/03.dr.ident_atac_clusters.png?raw=true" width="400"> | <img src="https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/03.coda_celltype.png?raw=true" width="400"> |


Therefore, in this sample, there are 3 epithelial cell types (Basal, Luminal secretory cells, Luminal hormone responsive cells), 1 immune cell types (myeloid cells), and 3 stromal cell types (fibroblast, endothelial cells, and perivascular cells). Interestingly, the LUmHR cells have 3 or 4 distinct clusters. You may be wondering why. 

**Spoiler Alert**: this is the cool part of our wellDA, revealing that there are diplolid and aneuploid cells from the same LumHR cell type. In addition, there are more than one subclones. Because of the dosage-effect of genomic copy number on expressions, different subclones will show different ATAC clusters. You will see the evidence in next [section](https://github.com/navinlabcode/wellDA-seq/blob/main/tutorial/04.wellDA_scCNA_annotation.md). 

<img src="https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/03.coda_dimplot4.png?raw=true" width="600">




