<!-- Written by: Yun Yan (https://github.com/Puriney) -->

# Global concordance of genotypes and epigenotypes

Full R script: <kbd>scripts/coda.annotate_DNA_lineage.s3_cna.public.R</kbd>

Prerequisite R scripts: run the following two scripts before running the full R script. 
- <kbd>scripts/copykit.medic2.polish.R</kbd>: Generate the MEDICC2 lineage tree and creates the consensus CNA profiles of subclones. 
- <kbd>scripts/std.coda.css_CNA_distance.public.R</kbd>: Calculate the genomic distance between subclones. 

**Input files:** 
- the analysis-ready folder of wellDA-seq data

**Rationale**: Within a tumor, there are multiple subclones showing different CNA profiles. Some subclones (e.g., C3 v.s. C4) consistently show different chromatin accessibility profiles -- indicating a global concordance between genotypes and epigenotypes. In contrast, some genetically divergent subclones (e.g., C1 v.s. C4) are epigenetically similar -- indicting a global discordance. This section aims to quantify this global concordance. 

<img src="https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/07.coda_crossdimplot.clones.png?raw=true" width="550">



## 1. Pairwise ATAC distance of subclones

The ATAC distances of subclones are not biased using some previously defined gene signatures. The calculation is unbiased and depends on the data itself by using the LSI components. Specifically, a consensus ATAC profile of a subclone is the median of the LSI components values of included single-cells. Last, the Euclidean distance matrix of these consensus ATAC profiles is computed, representing the pair-wise ATAC distances of the subclones. 

```R
#------ atac distance ------
## based on LSI components
DefaultAssay(obja) <- 'peaks'
n_lsi <- ncol(Embeddings(obja, reduction = 'lsi'))
p <- DepthCor(obja, n = n_lsi)
lsi_idx <- head(setdiff(1:n_lsi, which(abs(p$data$counts)>0.75)), 30)
atac_events <- Embeddings(obja, reduction = 'lsi')[, lsi_idx]

dict_cell2ident <- obja@meta.data[[ident_str]]; names(dict_cell2ident) <- Cells(obja)
atac_events <- apply(atac_events, 2, function(vv) {
  tapply(vv, dict_cell2ident, median)
})
atac_dist <- dist(atac_events)
atac_dist <- as.matrix(atac_dist)
atac_dist <- atac_dist[ident_vec, ident_vec]
write_csv(as.data.frame(atac_dist), 
          file.path(dir_res, 'atac_pairwise_distance.csv'))

```

## 2. Pairwise CNA distance of subclones

Run <kbd>scripts/copykit.medic2.polish.R</kbd> and <kbd>scripts/std.coda.css_CNA_distance.public.R</kbd> create the integer consensus CNA profiles of subclones and compute the Manhattan distances. 

<img src="https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/07.cna.heatmap_consensus_int.combo_coda.clones.png?raw=true" width="500">

Manhattan distance of the integer CNA profiles of subclones: 
```R
objd <- calcConsensus(objd, assay = 'integer', consensus_by = 'clones')
css_mat <- objd@consensus

```
```R
cna_dist_mat <- dist(t(css_mat), method = 'manhattan')
```

## 3. Calculate and visualize the concordance of genotypes and epigenotypes differences


To facilitate data visualization, we re-scale the value range of the ATAC distance `atac_dist` to the value range of the CNA distance `cna_dist_mat`. 

```R
atac_dist <- scales::rescale(
  x = as.matrix(as.dist(atac_dist)), 
  to = cna_dist_val_range) ## good idea to keep the scale similar to get a better visualization

atac_dist <- atac_dist[intersect(ident_vec, rownames(atac_dist)),intersect(ident_vec, rownames(atac_dist))]
```


Compute the Pearson correlation of the two distance matrices to represent the global concordance between genotypes and epigenotypes. 
```R
#------ concordance ------
stopifnot(all.equal(rownames(cna_dist_mat), rownames(atac_dist)))
stopifnot(all.equal(colnames(cna_dist_mat), colnames(atac_dist)))

score_GP_concordance_obj <- cor.test(
  as.dist(cna_dist_mat), 
  as.dist(atac_dist), method = 'pearson')

cat('score_GP_concordance = ', signif(score_GP_concordance, digits = 2), 'in sample', sample_name, '.\n')

```

In this sample, the global concordance is 0.43. 

```R
pdf(file.path(dir_res, 'heatmap.GP_concordance.cor.pdf'), 
      width = n_clones * 0.5  + 1, 
      height = n_clones *0.5 + 1, useDingbats = F)
(data.frame(CNA_dist=tmp_cna[lower.tri(tmp_cna)], 
            ATAC_dist=tmp_atac[lower.tri(tmp_atac)]) %>%
    ggplot(aes(x=CNA_dist, y=ATAC_dist)) + 
    geom_abline(slope = 1, intercept = 0, lty='dashed', color='grey') + 
    geom_point(pch=1, size=3) + 
    geom_smooth(method = 'lm', se=F) + 
    stat_cor() +
    coord_cartesian(xlim = cna_dist_val_range, ylim=cna_dist_val_range)
) %>% print()
dev.off()
```

<img src="https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/07.heatmap.GP_concordance.cor.png?raw=true" width="400">



Here we also visualize the two distance matrices, and the difference of the distance matrices to highlight which subclone pairs are most discordant. 

```R
tmp_cna <- cna_dist_mat[rev(genotree_tip_orders), rev(genotree_tip_orders)]
tmp_type <- 'CNA'
pt_cna <- Heatmap(
  matrix = tmp_cna, 
  name = sprintf('%s dist', tmp_type), 
  cluster_rows = F, cluster_columns = F, 
  col = circlize::colorRamp2(
    quantile(c(c(cna_dist_mat), c(atac_dist)), c(0.01, 0.99)), 
    hcl.colors(n=2, 'Blues 3', rev = F)),
  show_row_dend = T, show_column_dend = T, 
  width = unit(n_clones*0.5, 'inch'),
  height = unit(n_clones*0.5 /ncol(tmp_cna)*nrow(tmp_cna), 'inch'),
  row_names_side  = 'left',
  rect_gp = gpar(type = "none"), 
  cell_fun = function(j, i, x, y, width, height, fill) {
    if (i >= j) { 
      grid.rect(x, y, width, height, gp = gpar(fill = fill, col = fill))
      if (tmp_cna[i, j] > 0) {
        grid.text(sprintf("%d", tmp_cna[i, j]), x, y, gp = gpar(fontsize = 10, col='black'))
      }
    }
  },
  
  column_title = tmp_type, 
  column_title_gp = gpar(fontsize = 10), use_raster = F
)

is.wholenumber <- function(x, tol = .Machine$double.eps^0.5)  abs(x - round(x)) < tol

tmp_type <- 'ATAC'
tmp_atac <- atac_dist[rev(genotree_tip_orders), rev(genotree_tip_orders)]
pt_atac <- Heatmap(
  matrix = tmp_atac, 
  name = sprintf('%s dist', tmp_type), 
  cluster_rows = F, cluster_columns = F, 
  col = circlize::colorRamp2(
    quantile(c(c(cna_dist_mat), c(atac_dist)), c(0.01, 0.99)), 
    hcl.colors(n=2, 'Blues 3', rev = F)),
  show_row_dend = T, show_column_dend = T, 
  width = unit(n_clones*0.5, 'inch'),
  height = unit(n_clones*0.5 /ncol(tmp_atac)*nrow(tmp_atac), 'inch'),
  row_names_side  = 'left',
  rect_gp = gpar(type = "none"), 
  cell_fun = function(j, i, x, y, width, height, fill) {
    if (i >= j) {
      grid.rect(x, y, width, height, gp = gpar(fill = fill, col = fill))
      if (tmp_atac[i, j] > 0) {
        if (is.wholenumber(tmp_atac[i, j])) {
          grid.text(sprintf("%d", as.integer(tmp_atac[i, j])), x, y, gp = gpar(fontsize = 10, col='black'))
        } else {
          grid.text(sprintf("%.2f", tmp_atac[i, j]), x, y, gp = gpar(fontsize = 10, col='black'))
        }
      }
    }
  },
  column_title = tmp_type, 
  column_title_gp = gpar(fontsize = 10), use_raster = F
)
# pt_cna
# pt_atac
tmp_diff <- abs(tmp_cna - tmp_atac)
tmp_type <- 'abs diff'

pt_diff <- Heatmap(
  matrix = tmp_diff, 
  name = sprintf('%s', tmp_type), 
  cluster_rows = F, cluster_columns = F, 
  col = circlize::colorRamp2(
    c(cna_dist_val_range[1], mean(cna_dist_val_range), cna_dist_val_range[2]), 
    hcl.colors(n=3, 'Cyan-Magenta', rev = F)),
  show_row_dend = T, show_column_dend = T, 
  width = unit(n_clones*0.5, 'inch'),
  height = unit(n_clones*0.5 /ncol(tmp_diff)*nrow(tmp_diff), 'inch'),
  row_names_side  = 'left',
  rect_gp = gpar(type = "none"), 
  cell_fun = function(j, i, x, y, width, height, fill) {
    if (i >= j) {
      # only draw the low-triangle part
      grid.rect(x, y, width, height, gp = gpar(fill = fill, col = fill))
      if (tmp_diff[i, j] > 0) {
        if (is.wholenumber(tmp_diff[i, j])) {
          grid.text(sprintf("%d", as.integer(tmp_diff[i, j])), x, y, gp = gpar(fontsize = 10, col='black'))
        } else {
          grid.text(sprintf("%.2f", tmp_diff[i, j]), x, y, gp = gpar(fontsize = 10, col='black'))
        }
      }
    }      
  },
  column_title = tmp_type, 
  column_title_gp = gpar(fontsize = 10), use_raster = F
)

pdf(file.path(dir_res, 'heatmap.GP_concordance.pdf'), 
    width = n_clones * 0.8 *  3 + 1, 
    height = n_clones *0.8 + 2, useDingbats = F)
draw(
  pt_cna + pt_atac + pt_diff, 
  column_title = sprintf('concordance = %.2f | pval=%.3g', 
                         score_GP_concordance, 
                         score_GP_concordance_pval), 
  column_title_gp = gpar(fontsize = 10))
dev.off()

```

<img src="https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/07.heatmap.GP_concordance.png?raw=true" width="600">

In this sample, the subclones C2 and C4 had similar CNA profiles, however, they showed substantial difference at the epigenetic layer. 

