<!-- Written by: Yun Yan (https://github.com/Puriney) -->

# Identify the genetically heritable and plastic phenotypes

Full R script: 
- <kbd>scripts/coda.annotate_DNA_lineage.s3_cna.public.R</kbd>

Prerequisite R script:
- <kbd>scripts/std.cicero_inferRNA.R</kbd>: Infer RNA from ATAC. 
- <kbd>scripts/coda.annotate_DNA_lineage.s1_run_UCell.R</kbd>: Compute module scores of the known gene signatures. 
- <kbd>scripts/copykit.medic2.polish.R</kbd>: Generate the MEDICC2 lineage tree to facilitate data visualization.


**Rationale**: Previous studies found that aneuploid patterns in cancer cells are highly associated with genes controlling proliferation, with recurrently amplified genes promoting proliferation ([Davoli, T. et al. 2013](https://doi.org/10.1016/j.cell.2013.10.011)). In contrast, other tumorigenesis-related phenotypes, such as cGAS-STING signaling which is known for detecting cytosolic DNA and activating immune response, are less likely associated with specific genomic aberrations ([Ben-David, U. & Amon, A. 2020](https://doi.org/10.1038/s41576-019-0171-x)). Our hypothesis is that the genomic hardwired phenotypes are cancer cell-autonomous while the plastic phenotypes are non-autonomous.

## 1. Infer gene expression from the ATAC data

Because the ATAC data is sparse, we infer the gene expression using the ATAC data for the known tumorigenesis-related phenotypes from the MsigDB Cancer Hallmark ([link](https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp)). Run the R script <kbd>scripts/std.cicero_inferRNA.R</kbd>, which follows the Cicero's tutorial ([link](https://cole-trapnell-lab.github.io/cicero-release/docs_m3/#cicero-gene-activity-scores)). 

Each phenotype is a collection of genes, so we compute the 'module score' for the phenotypes (i.e., gene sets) by using the UCell package ([M. Andreatta & S. J. Carmona, 2021](https://doi.org/10.1016/j.csbj.2021.06.043)). Run the R script <kbd>scripts/coda.annotate_DNA_lineage.s1_run_UCell.R</kbd>. 


## 2. Quantify the level of heritability or plasticity

We estimate whether the pair-wise genomic distances of subclones are correlated with the phenotypic differences of the subclones such that the phenotype is defined by a specific gene signature. Therefore, for each gene signature, we compute a score evaluating whether the its expression  in the subclones is associated with the phylogenetic tree. 


We focused the analysis on the gene signatures showing significant expression among subclones. 
```R
message('Load AddModuleScore ', module_method)
df_signature_res <- read_rds(file.path(dir_modulescore, 'addmodulescore.df.rds'))
phenotypes_opts <- colnames(df_signature_res)

for (i in seq_along(phenotypes_opts)) {
  stat.test <- df_meta %>% kruskal_test( as.formula(sprintf('%s~%s', phenotypes_opts[i], z)))
  pheno_ks_test_pval_vec <- c(pheno_ks_test_pval_vec, stat.test$p)
}; names(pheno_ks_test_pval_vec) <- phenotypes_opts
pheno_ks_test_qval_vec <- p.adjust(pheno_ks_test_pval_vec, method = 'BH')
## pval < 0.05
phenotypes_sig_diverse_opts <- phenotypes_opts[which(pheno_ks_test_pval_vec < 0.05)]
if (length(phenotypes_sig_diverse_opts) <= 2) {
  cat('[warn] there are only ', length(phenotypes_sig_diverse_opts), 
      ' sig diff features in sample', sample_name, '.\n' )
}

```


For each gene signature, compute the pair-wise distances of subclones. Then compare this distance matrix with the pre-computed CNA distances of subclones by the Pearson correlation. If the correlation is as high as 1, it indicates the gene signature is tightly bound to the tumor evolution. If the correlation is low, this gene signature has little correlation with the subclone evolution, meaning this is a plastic gene signature. 

```R
calc_heritability_score_orig <- function(d1, d2) {
  # d1, d2: two distance objects
  lb <- labels(d1)
  m2 <- as.matrix(d2)
  m2 <- m2[lb, lb]
  d2 <- as.dist(m2)
  o <- cor.test(d1, d2, method = 'pearson')  # similar as spearman
  c('r' = as.numeric(o$estimate), 
    'pval' = as.numeric(o$p.value))
}


phenotypes_sig_diverse_heritability_orig_test <- sapply(
  phenotypes_sig_diverse_opts, function(xx) {
    calc_heritability_score_orig(
      as.dist(cna_dist_mat),
      dist(mat_pheno_viz_zscore[, xx, drop=F]) )
  }
)

phenotypes_sig_diverse_heritability_orig <- phenotypes_sig_diverse_heritability_orig_test['r', ]

```

In the script, we have two other calculation strategies: Geary's C (see [wikipedia](https://en.wikipedia.org/wiki/Geary%27s_C))and PATH ([J. S. Schiffman 2023](https://github.com/landau-lab/PATH)). But we decide not using them because:
1) the Geary's C metric is more likely to work well with spatial coordinates data, although the genomic distance between the subclones can be used to assign the weight $w$ in computing Geary's C. 
2) the PATH metric is more likely to work well with the single-cell point-mutation data. 


## 3. Visualize the association of gene expressions and CNAs

Create a data frame where each row is a gene signature and the column saves the heritability score. 
```R

report_heritability <- data.frame(
  phenotype = phenotypes_sig_diverse_opts, 
  heritability_cor = phenotypes_sig_diverse_heritability_orig[phenotypes_sig_diverse_opts], 
  heritability_cor_pval = phenotypes_sig_diverse_heritability_orig_pval[phenotypes_sig_diverse_opts], 
  heritability_path = phenotypes_sig_diverse_xcor_scores[phenotypes_sig_diverse_opts], 
  heritability_gearyc = phenotypes_sig_diverse_heritability_gearyc[phenotypes_sig_diverse_opts], 
  stringsAsFactors = F)
```


Create the genetic phylogenetic tree by MEDICC2. Run <kbd>scripts/copykit.medic2.polish.R</kbd>. To clarify, MEDICC2 is used here just for visualization. The genomic distance between subclones are is calculated using the MEDICC2 result or the tree-based distance (e.g., Cophenetic distance). Because the original CNA profiles have been the ground truth, the MEDICC2-based and tree-based distances all simplified the tree, resulting an approximated estimation of genetic distance between subclones.


Run the R script <kbd>scripts/coda.annotate_DNA_lineage.s3_cna.public.R</kbd> to generate this data visualization: 

<img src="https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/08.cna.heatmap_consensus_int.combo_coda.clones.png?raw=true" width="600">

<img src="https://github.com/navinlabcode/wellDA-seq/blob/main/website_resource/tutorial/08.concordance.heritability.medic2.msigdb_hallmark.zscore.birdview_SigOnly.combo.png?raw=true" width="600">


In the tumor evolution tree, C2 is the most early subclone and C1 the most descent subclone. The reactive oxygen species pathway has the high expression in C2 an C4, and has the low expression in C1 and C3, therefore, its expression pattern is correlated with the evolution tree, making it a 'heritable' phenotype. 

In contrast, the complement gene signature has high expressions in C1 and C4, despite their distinct genomic profiles, making it a 'plastic' phenotype. 


